Document errors implemented in 2025.01 (#243)
neo4j/docs-status-codes: modules/ROOT/pages/errors/gql-errors.adoc

// Status description:: error: data exception - index required by existing constraint. A requested operation can not be performed on the specified index because the index is part of a constraint.

=== 22N73

Status description:: error: data exception - constraint conflicts with existing index. Constraint conflicts with already existing index `{ $idx }`.

=== 22N74

> Status description:: error: data exception - index conflicts with existing constraint. An index that belongs to the constraint `{ $constr }` contains a conflicting index.

Author: This sounds odd. I guess we are trying to say
```suggestion
Status description:: error: data exception - index conflicts with existing constraint. Index conflict with an already existing index that belongs to the constraint `{ $constr }`.
```

> Status description:: error: data exception - index conflicts with existing constraint. An index that belongs to the constraint `{ $constr }` contains a conflicting index.

Reviewer: As Satia were on vacation, I made a card for this one, so we can fix it later. Because if we fix it in the codebase, that will not be in 2025.01 anyway

Status description:: error: syntax error or access rule violation - invalid use of `MATCH`. 'MATCH ...' cannot directly follow an 'OPTIONAL MATCH ...'. Use a `WITH` clause between them.

=== 42I34

Status description:: error: syntax error or access rule violation - invalid use of pattern expression. A pattern expression can only be used to test the existence of a pattern. Use a pattern comprehension instead.

=== 42I36

> Status description:: error: syntax error or access rule violation - invalid use of `REPORT STATUS`. 'REPORT STATUS' can only be used when specifying 'ON ERROR CONTINUE' or 'ON ERROR BREAK'.

Author: Yes, we need to discuss it and make it consistent in both docs and codebase. It might make sense to leave the backticks for the docs only unless something needs to be explicitly quoted in the codebase.

Status description:: error: syntax error or access rule violation - invalid parameter. Invalid parameter `{ $param }`.

=== 42N53

> Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode 'REPEATABLE ELEMENTS` Use a path selector or add an upper bound to the quantified path pattern.

Reviewer 2: Agree with @Lojjs. Also it turns out that, for now at least, we don't think it's safe to allow just path selectors under repeatable elements. So there is only one fix, adding an upper bound to any quantified path pattern. 

Here is my suggested change:

```suggestion
Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode 'REPEATABLE ELEMENTS'. Add an upper bound to the quantified path pattern.
```

Reviewer: @renetapopova If you change it like this (also remove the path selector part) here in docs, I can fix it in the spreadsheet and codebase. I anyway plan to implement the last error code for planner today so can fix this in the same PR.

Reviewer: Updated in the codebase here https://github.com/neo-technology/neo4j/pull/29116

Reviewer: @NataliaIvakina FYI, this commented is still unhandled (or rather Reneta only fixed my complaint and not Finbar's)

> Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode 'REPEATABLE ELEMENTS` Use a path selector or add an upper bound to the quantified path pattern.

Author: I agree with you. I'll update it here.

Status description:: error: syntax error or access rule violation - invalid parameter. Invalid parameter `{ $param }`.

=== 42N53

> Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode `REPEATABLE ELEMENTS'. Add an upper bound to the quantified path pattern.

Reviewer 3: ```suggestion
Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode `REPEATABLE ELEMENTS`. Add an upper bound to the quantified path pattern.
```

> Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode `REPEATABLE ELEMENTS`. Use a path selector or add an upper bound to the quantified path pattern.

Reviewer 3: ```suggestion
Status description:: error: syntax error or access rule violation - unsafe usage of repeatable elements. The quantified path pattern may yield an infinite number of rows under match mode `REPEATABLE ELEMENTS'. Add an upper bound to the quantified path pattern.
```

Reviewer: Doesn't this become wrong because now again REPEATABLE ELEMENTS have different quote before and after. I'm fine with either quoting for now but if they are different I suspect it will screw up docs formatting

Reviewer 3: My bad. I fixed backticks

Status description:: error: system configuration or operation exception - unsupported operation of a sharded database. Feature: `{ $feat }` is not available in a sharded database.

=== 51N72

> Status description:: error: system configuration or operation exception - memory pool out of memory. Failed to allocate memory in a memory pool. See `{ $cfgSetting }` in the neo4j configuration.

Reviewer: ```suggestion
Status description:: error: system configuration or operation exception - memory pool out of memory. Failed to allocate memory in a memory pool. See `{ $cfgSetting }` in the neo4j.conf file.
```

I think this and the two below were changed based on PR feedback after you made this PR.

=== 51N73

> Status description:: error: system configuration or operation exception - transaction memory limit reached. The transaction used more memory than was allowed. The maximum allowed size for a transaction can be configured with `{ $cfgSetting }` in the neo4j configuration.

Reviewer: ```suggestion
Status description:: error: system configuration or operation exception - transaction memory limit reached. The transaction used more memory than was allowed. The maximum allowed size for a transaction can be configured with `{ $cfgSetting }` in the neo4j.conf file.
```

=== 51N74

> Status description:: error: system configuration or operation exception - maximum number of transactions reached. Failed to start a new transaction. The limit of concurrent transactions is reached. Increase the number of concurrent transactions using `{ $cfgSetting }` in the neo4j configuration.

Reviewer: ```suggestion
Status description:: error: system configuration or operation exception - maximum number of transactions reached. Failed to start a new transaction. The limit of concurrent transactions is reached. Increase the number of concurrent transactions using `{ $cfgSetting }` in the neo4j.conf file.
```

Status description:: error: procedure exception - no such seeder. The specified seeding server with id `{ $server }` was not found. Verify that the spelling is correct.

=== 52N22

> Status description:: error: procedure exception - invalid procedure argument. Invalid argument `{ $value }` for `{ $procParam }` on procedure `{ $proc }`. The expected format of `{ $procParam }` is `{ $procParamFmt }`.

Reviewer: ```suggestion
Status description:: error: procedure exception - invalid procedure argument. Invalid argument `{ $field }` for `{ $procParam }` on procedure `{ $proc }`. The expected format of `{ $procParam }` is `{ $procParamFmt }`.
```

This one was inconsistent between spreadsheet and codebase, I have updated the spreadsheet now

Status description:: error: procedure exception - procedure compilation failed. Failed to compile procedure/function defined in `{ $procClass }`: `{ $msg }`

=== 52N36

> Status description:: error: procedure exception - invalid procedure argument with API documentation hint. Invalid argument `{ $value }` for `{ $procParam }` on procedure `{ $proc }`. The expected format of `{ $procParam }` is outlined in the `{ $procParamFmt }` API documentation.

Reviewer: ```suggestion
Status description:: error: procedure exception - invalid procedure argument with API documentation hint. Invalid argument `{ $field }` for `{ $procParam }` on procedure `{ $proc }`. The expected format of `{ $procParam }` is outlined in the `{ $procParamFmt }` API documentation.
```

This one was different in spreadsheet and implementation, I have updated the spreadsheet.

=== 52N37

Status description:: error: procedure exception - procedure execution error. Execution of the procedure `{ $proc } ` failed.

=== 52U00

Status description:: error: procedure exception - custom procedure execution error cause. Execution of the procedure `{ $proc }` failed due to `{ $msgTitle }`: { $msg }

> 

Reviewer: Maybe it does not matter as you are anyway doing a refactoring in one of your other PRs but in this format we should start a new subsection for Function exception here

Author: You're right. I'll add it here as well.

> === 53N34

Author: ```suggestion
== Function exceptions

=== 53N34
```

Status description:: error: procedure exception - procedure compilation failed. Failed to compile procedure/function defined in `{ $procClass }`: `{ $msg }`

=== 52N36

Status description:: error: procedure exception - invalid procedure argument with API documentation hint. Invalid argument `{ $field }` for `{ $procParam }` on procedure `{ $proc }`. The expected format of `{ $procParam }` is outlined in the `{ $procParamFmt }` API documentation.

=== 52N37

Status description:: error: procedure exception - procedure execution error. Execution of the procedure `{ $proc } ` failed.

=== 52U00

> Status description:: error: procedure exception - custom procedure execution error cause. Execution of the procedure `{ $proc }` failed due to `{ $msgTitle }`: { $msg }

Author: ```suggestion
Status description:: error: procedure exception - custom procedure execution error cause. Execution of the procedure `{ $proc }` failed due to `{ $msgTitle }`: { $msg }.
```

== Function exceptions

=== 53N34

> Status description:: error: function exception - function restricted. `{ $fun }` is is restricted and accesses database internals. User-defined function restriction is controlled by the dbms.security.procedures.unrestricted setting. Only un-restrict user-defined functions you can trust with access to database internals.

Author: ```suggestion
Status description:: error: function exception - function restricted. `{ $fun }` is restricted and accesses database internals. User-defined function restriction is controlled by the `dbms.security.procedures.unrestricted` setting. Only un-restrict user-defined functions you can trust with access to database internals.
```

=== 53N35

Status description:: error: function exception - function compilation failed. Failed to compile function defined in `{ $funClass }`: { $msg }

=== 53N37

Status description:: error: function exception - function execution error. Execution of the function `{ $fun }` failed.

=== 53U00

> Status description:: error: function exception - custom function execution error cause. Execution of the function `{ $fun }` failed due to `{ $msgTitle }`: { $msg }

Author: ```suggestion
Status description:: error: function exception - custom function execution error cause. Execution of the function `{ $fun }` failed due to `{ $msgTitle }`: { $msg }.
```
